Implement save password working on Linux

### Basic Info

- I'm running Cryptomator on: any Linux
- I'm using Cryptomator in version: 1.2.3

### Description

Save password is currently not implemented on Linux systems. Due to differences in the distributions this may not be solvable by using a single API but several interfaces, like Gnome Keyring, will have to be supported.

First step is to investigate which interfaces are the most common (would suggest we should aim at 95% market share) and implement them.

For now the following shall be supported:
- [ ] GNOME Keyring
- [ ] more may follow

***Note*** Extracted this issue from #40. This is a separate feature which can be used on its own without #40 being implemented.

> ```
> java.lang.ExceptionInInitializerError: null
> 	at org.freedesktop.dbus.connections.Transport.connect(Transport.java:119)
> 	at org.freedesktop.dbus.connections.Transport.<init>(Transport.java:70)
> 	at org.freedesktop.dbus.connections.AbstractConnection.<init>(AbstractConnection.java:149)
> 	at org.freedesktop.dbus.connections.impl.DBusConnection.<init>(DBusConnection.java:284)
> 	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:122)
> 	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:142)
> 	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:200)
> 	at org.freedesktop.dbus.connections.impl.DBusConnection.getConnection(DBusConnection.java:157)
> 	at org.freedesktop.secret.TransportEncryption.<init>(TransportEncryption.java:36)
> 	at org.freedesktop.secret.simple.SimpleCollection.init(SimpleCollection.java:112)
> 	at org.freedesktop.secret.simple.SimpleCollection.<init>(SimpleCollection.java:44)
> 	at org.cryptomator.keychain.LinuxSecretServiceAccess.<init>(LinuxSecretServiceAccess.java:20)
> Caused by: java.lang.RuntimeException: Could not load library from any given source: [SYSTEM_PATH, CUSTOM_PATH, CLASS_PATH]
> 	at com.github.hypfvieh.system.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:89)
> 	at com.github.hypfvieh.system.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:69)
> 	at cx.ath.matthew.unix.UnixSocket.<clinit>(UnixSocket.java:54)
> ```

I was taking a look, but have some problems to get that exception thrown.
What is confusing me too is that running Cryptomator on Windows (even the default develop branch) gives me an

```12:57:58.862 [JavaFX Application Thread] ERROR org.cryptomator.jni.JniModule - Could not load JNI lib WinFunctions.dll from path C:\Program Files\Java\jdk-12.0.1\bin;C:\WINDOWS\Sun\Java\bin;C:\WINDOWS\system32;C:\WINDOWS;C:\Program Files (x86)\Parallels\Parallels Tools\Applications;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Git\cmd;C:\Program Files (x86)\Gpg4win\..\GnuPG\bin;C:\Users\ralph\AppData\Local\Microsoft\WindowsApps;C:\Program Files\Java\jdk-12.0.1\bin;C:\Users\ralph\Runnable\apache-maven-3.6.1\bin;;.```

I am compiling Cryptomator for the first time on Windows and probably there is something wrong with my configuration. But with this setting I doubt that I can do a meaningful test.

A suggestion: what about setting the keyring to null again in case of problems?
Does this change anything to the better?
```
public LinuxSecretServiceAccess() {
		SimpleCollection keyring = null;
		try {
			keyring = new SimpleCollection();
		} catch (Exception e) {
			// Accessing secret-service DBus API failed
			keyring = null;  // <-------
		} finally {
			gnomeLoginKeyring = Optional.ofNullable(keyring);
		}
}
```

> Also, I was not able to create multiple jars for the _keychain_ module with different classifiers. This results in larger Mac and Windows builds due to the included unnecessary (transitive) dependencies.

Maybe the maven copy-dependencies plugin can be configured to copy the secret-service.jar only to linux-libs?

> WinFunctions is not compiled via Maven. We've done it manually and added the binary to [cryptomator/cryptomator-win](https://github.com/cryptomator/cryptomator-win). When launching from your IDE, it is not present and save password will not work. You can of course download the dll to your working directory manually, if you want to test save pw on Windows.
> 

Thanks for explaining this.

I also was not able to get the transitive dependencies to work. `includeGroupIds` and `includeArtifactIds` in the buildkit pom did not work as expected.

A different approach I tried was to exclude the generated secret-service.jar from the Windows and the macOS build.
https://github.com/purejava/cryptomator/commit/1df87cc23f98f8e2341d77dc6967ecbe3d45576a

I installed the tools required on Windows to generate the executable and started it, but this brings us back to the other problem: the class does not exist in this case.
```
14:26:03.396 [main] INFO  org.cryptomator.launcher.Cryptomator - Starting Cryptomator 1.5.0-SNAPSHOT on Windows 10 10.0 (amd64)
14:26:03.537 [main] INFO  o.c.launcher.IpcProtocolImpl - Received launch args: 
14:26:03.840 [JavaFX Application Thread] INFO  org.cryptomator.launcher.Cryptomator - JavaFX application started.
14:26:04.040 [JavaFX Application Thread] INFO  org.cryptomator.jni.JniModule - loaded WinFunctions.dll
14:26:04.043 [JavaFX Application Thread] INFO  org.cryptomator.launcher.Cryptomator - JavaFX application stopped.
14:26:04.050 [main] ERROR org.cryptomator.launcher.Cryptomator - Terminating due to error
java.lang.RuntimeException: Exception in Application start method
	at com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:900)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.base/java.lang.Thread.run(Thread.java:835)
Caused by: java.lang.NoClassDefFoundError: org/freedesktop/secret/simple/SimpleCollection
	at org.cryptomator.keychain.LinuxSecretServiceAccess.<init>(LinuxSecretServiceAccess.java:20)
	at org.cryptomator.keychain.LinuxSecretServiceAccess_Factory.get(LinuxSecretServiceAccess_Factory.java:16)
	at org.cryptomator.keychain.LinuxSecretServiceAccess_Factory.get(LinuxSecretServiceAccess_Factory.java:6)
	at org.cryptomator.keychain.KeychainModule_ProvideKeychainAccessStrategiesFactory.get(KeychainModule_ProvideKeychainAccessStrategiesFactory.java:37)
	at org.cryptomator.keychain.KeychainModule_ProvideKeychainAccessStrategiesFactory.get(KeychainModule_ProvideKeychainAccessStrategiesFactory.java:9)
	at dagger.internal.SetFactory.get(SetFactory.java:119)
	at dagger.internal.SetFactory.get(SetFactory.java:37)
	at org.cryptomator.keychain.KeychainModule_ProvideSupportedKeychainFactory.get(KeychainModule_ProvideSupportedKeychainFactory.java:29)
	at org.cryptomator.keychain.KeychainModule_ProvideSupportedKeychainFactory.get(KeychainModule_ProvideSupportedKeychainFactory.java:10)
	at org.cryptomator.ui.model.AutoUnlocker_Factory.get(AutoUnlocker_Factory.java:33)
	at org.cryptomator.ui.model.AutoUnlocker_Factory.get(AutoUnlocker_Factory.java:10)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at org.cryptomator.ui.controllers.MainController_Factory.get(MainController_Factory.java:77)
	at org.cryptomator.ui.controllers.MainController_Factory.get(MainController_Factory.java:17)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at org.cryptomator.ui.controllers.ViewControllerModule_ProvideMainControllerFactory.get(ViewControllerModule_ProvideMainControllerFactory.java:26)
	at org.cryptomator.ui.controllers.ViewControllerModule_ProvideMainControllerFactory.get(ViewControllerModule_ProvideMainControllerFactory.java:8)
	at org.cryptomator.ui.controllers.ViewControllerLoader.constructController(ViewControllerLoader.java:48)
	at javafx.fxml.FXMLLoader$ValueElement.processAttribute(FXMLLoader.java:938)
	at javafx.fxml.FXMLLoader$InstanceDeclarationElement.processAttribute(FXMLLoader.java:980)
	at javafx.fxml.FXMLLoader$Element.processStartElement(FXMLLoader.java:227)
	at javafx.fxml.FXMLLoader$ValueElement.processStartElement(FXMLLoader.java:752)
	at javafx.fxml.FXMLLoader.processStartElement(FXMLLoader.java:2722)
	at javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2552)
	at javafx.fxml.FXMLLoader.load(FXMLLoader.java:2450)
	at org.cryptomator.ui.controllers.ViewControllerLoader.load(ViewControllerLoader.java:36)
	at org.cryptomator.launcher.Cryptomator$MainApp.start(Cryptomator.java:109)
	at com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:846)
	at com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:455)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:389)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
	... 1 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.freedesktop.secret.simple.SimpleCollection
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	... 36 common frames omitted
14:26:04.050 [Thread-0] INFO  o.c.launcher.CleanShutdownPerformer - Goodbye.
```

So we need some similar mechanism for Linux as we have it for Windows and macOS, where the platform specific code is loaded as a library.

--- a/main/buildkit/assembly-linux.xml
+++ b/main/buildkit/assembly-linux.xml
@@ -28,5 +28,12 @@
 			</includes>
 			<outputDirectory>libs</outputDirectory>
 		</fileSet>
+		<fileSet>
+			<directory>target/linux-secret-service</directory>
+			<includes>
+				<include>*.jar</include>
+			</includes>
+			<outputDirectory>libs</outputDirectory>
+		</fileSet>
 	</fileSets>
 </assembly>

--- a/main/buildkit/pom.xml
+++ b/main/buildkit/pom.xml
@@ -62,6 +62,7 @@
 						<configuration>
 							<outputDirectory>${project.build.directory}/libs</outputDirectory>
 							<excludeClassifiers>linux,mac,win</excludeClassifiers>
+							<excludeArtifactIds>dbus-java,secret-service,hkdf,java-utils</excludeArtifactIds>
 						</configuration>
 					</execution>
 					<execution>
@@ -74,6 +75,17 @@
 							<outputDirectory>${project.build.directory}/linux-libs</outputDirectory>
 							<includeGroupIds>org.openjfx</includeGroupIds>
 							<classifier>linux</classifier>
+						</configuration>
+					</execution>
+					<execution>
+						<id>copy-linux-secret-service</id>
+						<phase>prepare-package</phase>
+						<goals>
+							<goal>copy-dependencies</goal>
+						</goals>
+						<configuration>
+							<outputDirectory>${project.build.directory}/linux-secret-service</outputDirectory>
+							<includeArtifactIds>dbus-java,secret-service,hkdf,java-utils</includeArtifactIds>
 						</configuration>
 					</execution>
 					<execution>

--- a/main/keychain/pom.xml
+++ b/main/keychain/pom.xml
@@ -39,7 +39,14 @@
 			<groupId>com.google.dagger</groupId>
 			<artifactId>dagger</artifactId>
 		</dependency>
-		
+
+		<!-- secret-service lib -->
+		<dependency>
+			<groupId>de.swiesend</groupId>
+			<artifactId>secret-service</artifactId>
+			<version>1.0.0-RC.3</version>
+		</dependency>
+
 		<!-- Logging -->
 		<dependency>
 			<groupId>org.slf4j</groupId>

--- a/main/keychain/src/main/java/org/cryptomator/keychain/GnomeKeyringAccess.java
+++ b/main/keychain/src/main/java/org/cryptomator/keychain/GnomeKeyringAccess.java
@@ -1,0 +1,9 @@
+package org.cryptomator.keychain;
+
+public interface GnomeKeyringAccess {
+	public void storePassword(String key, CharSequence passphrase);
+
+	public char[] loadPassword(String key);
+
+	public void deletePassword(String key);
+}

--- a/main/keychain/src/main/java/org/cryptomator/keychain/GnomeKeyringAccessImpl.java
+++ b/main/keychain/src/main/java/org/cryptomator/keychain/GnomeKeyringAccessImpl.java
@@ -1,0 +1,52 @@
+package org.cryptomator.keychain;
+
+import org.freedesktop.secret.simple.SimpleCollection;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GnomeKeyringAccessImpl implements GnomeKeyringAccess {
+	private static final Logger LOG = LoggerFactory.getLogger(GnomeKeyringAccessImpl.class);
+	private SimpleCollection keyring;
+
+	public GnomeKeyringAccessImpl() {
+		try {
+			keyring = new SimpleCollection();
+		} catch (IOException e) {
+			LOG.error("D-Bus reports a problem.", e);
+		}
+	}
+
+	public void storePassword(String key, CharSequence passphrase) {
+		List<String> list = keyring.getItems(createAttributes(key));
+		if (list == null) {
+			keyring.createItem("Cryptomator", passphrase, createAttributes(key));
+		}
+	}
+
+	public char[] loadPassword(String key) {
+		List<String> list = keyring.getItems(createAttributes(key));
+		if (list != null) {
+			return keyring.getSecret(list.get(0));
+		} else {
+			return null;
+		}
+	}
+
+	public void deletePassword(String key) {
+		List<String> list = keyring.getItems(createAttributes(key));
+		if (list != null) {
+			keyring.deleteItem(list.get(0));
+		}
+	}
+
+	private Map<String, String> createAttributes(String key) {
+		Map<String, String> attributes = new HashMap();
+		attributes.put("Vault", key);
+		return attributes;
+	}
+}

--- a/main/keychain/src/main/java/org/cryptomator/keychain/KeychainModule.java
+++ b/main/keychain/src/main/java/org/cryptomator/keychain/KeychainModule.java
@@ -31,8 +31,8 @@ public class KeychainModule {
 
 	@Provides
 	@ElementsIntoSet
-	Set<KeychainAccessStrategy> provideKeychainAccessStrategies(MacSystemKeychainAccess macKeychain, WindowsProtectedKeychainAccess winKeychain) {
-		return Sets.newHashSet(macKeychain, winKeychain);
+	Set<KeychainAccessStrategy> provideKeychainAccessStrategies(MacSystemKeychainAccess macKeychain, WindowsProtectedKeychainAccess winKeychain, LinuxSecretServiceAccess linKeychain) {
+		return Sets.newHashSet(macKeychain, winKeychain, linKeychain);
 	}
 
 	@Provides

--- a/main/keychain/src/main/java/org/cryptomator/keychain/LinuxKeychainTester.java
+++ b/main/keychain/src/main/java/org/cryptomator/keychain/LinuxKeychainTester.java
@@ -1,0 +1,25 @@
+package org.cryptomator.keychain;
+
+import java.util.Optional;
+
+public class LinuxKeychainTester {
+	public static boolean secretServiceIsAvailable() {
+		try {
+			Class.forName("org.freedesktop.secret.simple.SimpleCollection");
+			return true;
+		} catch (ClassNotFoundException e) {
+			return false;
+		}
+	}
+
+	public static Optional<GnomeKeyringAccess> getSecretService() {
+		if (!secretServiceIsAvailable()) return Optional.empty();
+		try {
+			Class clazz = Class.forName("org.cryptomator.keychain.GnomeKeyringAccessImpl");
+			GnomeKeyringAccess keyring = (GnomeKeyringAccess) clazz.getDeclaredConstructor().newInstance();
+			return Optional.of(keyring);
+		} catch (Exception e) {
+			return Optional.empty();
+		}
+	}
+}

--- a/main/keychain/src/main/java/org/cryptomator/keychain/LinuxSecretServiceAccess.java
+++ b/main/keychain/src/main/java/org/cryptomator/keychain/LinuxSecretServiceAccess.java
@@ -2,66 +2,38 @@ package org.cryptomator.keychain;
 
 import com.google.common.base.Preconditions;
 import org.apache.commons.lang3.SystemUtils;
-import org.freedesktop.secret.simple.SimpleCollection;
 
 import javax.inject.Inject;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
 import java.util.Optional;
 
 public class LinuxSecretServiceAccess implements KeychainAccessStrategy {
-	private final Optional<SimpleCollection> gnomeLoginKeyring;
+	private final Optional<GnomeKeyringAccess> gnomeLoginKeyring;
 
 	@Inject
 	public LinuxSecretServiceAccess() {
-		SimpleCollection keyring = null;
-		try {
-			keyring = new SimpleCollection();
-		} catch (Exception e) {
-			// Accessing secret-service DBus API failed
-		} finally {
-			gnomeLoginKeyring = Optional.ofNullable(keyring);
-		}
+		gnomeLoginKeyring = LinuxKeychainTester.getSecretService();
 	}
 
 	@Override
 	public boolean isSupported() {
-		return SystemUtils.IS_OS_LINUX && gnomeLoginKeyring.isPresent();
+		return SystemUtils.IS_OS_LINUX && LinuxKeychainTester.getSecretService().isPresent();
 	}
 
 	@Override
 	public void storePassphrase(String key, CharSequence passphrase) {
 		Preconditions.checkState(gnomeLoginKeyring.isPresent());
-		List<String> list = gnomeLoginKeyring.get().getItems(createAttributes(key));
-		if (list == null) {
-			gnomeLoginKeyring.get().createItem("Cryptomator", passphrase, createAttributes(key));
-		}
+		gnomeLoginKeyring.get().storePassword(key, passphrase);
 	}
 
 	@Override
 	public char[] loadPassphrase(String key) {
 		Preconditions.checkState(gnomeLoginKeyring.isPresent());
-		List<String> list = gnomeLoginKeyring.get().getItems(createAttributes(key));
-		if (list != null) {
-			return gnomeLoginKeyring.get().getSecret(list.get(0));
-		} else {
-			return null;
-		}
+		return gnomeLoginKeyring.get().loadPassword(key);
 	}
 
 	@Override
 	public void deletePassphrase(String key) {
 		Preconditions.checkState(gnomeLoginKeyring.isPresent());
-		List<String> list = gnomeLoginKeyring.get().getItems(createAttributes(key));
-		if (list != null) {
-			gnomeLoginKeyring.get().deleteItem(list.get(0));
-		}
-	}
-
-	private Map<String, String> createAttributes(String key) {
-		Map<String, String> attributes = new HashMap();
-		attributes.put("Vault", key);
-		return attributes;
+		gnomeLoginKeyring.get().deletePassword(key);
 	}
 }

--- a/main/keychain/src/main/java/org/cryptomator/keychain/LinuxSecretServiceAccess.java
+++ b/main/keychain/src/main/java/org/cryptomator/keychain/LinuxSecretServiceAccess.java
@@ -1,0 +1,67 @@
+package org.cryptomator.keychain;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang3.SystemUtils;
+import org.freedesktop.secret.simple.SimpleCollection;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class LinuxSecretServiceAccess implements KeychainAccessStrategy {
+	private final Optional<SimpleCollection> gnomeLoginKeyring;
+
+	@Inject
+	public LinuxSecretServiceAccess() {
+		SimpleCollection keyring = null;
+		try {
+			keyring = new SimpleCollection();
+		} catch (Exception e) {
+			// Accessing secret-service DBus API failed
+		} finally {
+			gnomeLoginKeyring = Optional.ofNullable(keyring);
+		}
+	}
+
+	@Override
+	public boolean isSupported() {
+		return SystemUtils.IS_OS_LINUX && gnomeLoginKeyring.isPresent();
+	}
+
+	@Override
+	public void storePassphrase(String key, CharSequence passphrase) {
+		Preconditions.checkState(gnomeLoginKeyring.isPresent());
+		List<String> list = gnomeLoginKeyring.get().getItems(createAttributes(key));
+		if (list == null) {
+			gnomeLoginKeyring.get().createItem("Cryptomator", passphrase, createAttributes(key));
+		}
+	}
+
+	@Override
+	public char[] loadPassphrase(String key) {
+		Preconditions.checkState(gnomeLoginKeyring.isPresent());
+		List<String> list = gnomeLoginKeyring.get().getItems(createAttributes(key));
+		if (list != null) {
+			return gnomeLoginKeyring.get().getSecret(list.get(0));
+		} else {
+			return null;
+		}
+	}
+
+	@Override
+	public void deletePassphrase(String key) {
+		Preconditions.checkState(gnomeLoginKeyring.isPresent());
+		List<String> list = gnomeLoginKeyring.get().getItems(createAttributes(key));
+		if (list != null) {
+			gnomeLoginKeyring.get().deleteItem(list.get(0));
+		}
+	}
+
+	private Map<String, String> createAttributes(String key) {
+		Map<String, String> attributes = new HashMap();
+		attributes.put("Vault", key);
+		return attributes;
+	}
+}

--- a/main/keychain/src/test/java/org/cryptomator/keychain/TestKeychainModule.java
+++ b/main/keychain/src/test/java/org/cryptomator/keychain/TestKeychainModule.java
@@ -12,7 +12,7 @@ import com.google.common.collect.Sets;
 public class TestKeychainModule extends KeychainModule {
 
 	@Override
-	Set<KeychainAccessStrategy> provideKeychainAccessStrategies(MacSystemKeychainAccess macKeychain, WindowsProtectedKeychainAccess winKeychain) {
+	Set<KeychainAccessStrategy> provideKeychainAccessStrategies(MacSystemKeychainAccess macKeychain, WindowsProtectedKeychainAccess winKeychain, LinuxSecretServiceAccess linKeychain) {
 		return Sets.newHashSet(new MapKeychainAccess());
 	}
 

